Embedded bots UI [Update on #5811]

roberthoenig opened the pull request:
This PR is based on @abhijeetkaur 's PR for adding embedded bots to the UI.

### How does this differ from #5811?

I didn't modify Abhijeet's commits, but added two code clean-up commits on top of them. Additionally, I added a commit to fetch the embedded bots list automatically instead of hard-coding it in the template.

The original PR contains the request:
> Can we display the name of the specific embedded bot being used somewhere here? I'd consider putting it in the "bot type" row: e.g. "Embedded bot: converter". Might need to do some tweaking to the bots flow to pass that through the system, but seems like it'd be helpful.

From what I've seen in the code path, we need to either add some database field to bot users, or make some query to the service table to implement this. I think this could go hand in hand with https://github.com/zulip/zulip/pull/5665

### Why is this WIP?

* While an added embedded bot is responsive, there seems to be some problem with the message handling, which I still need to look into. Once that's fixed, all commits but `Add view to fetch embedded bots for user selection.` should be good to merge.

@timabbott pinging you because this is WIP, and it would be good to have some feedback if this is the right direction (also in the context of other PRs for embedded bots like state handling, etc.).

derAnfaenger commented:
@timabbott once the tests pass this should be ready to merge, if you agree with adding additional commit on top of abhijeet's work instead of rebasing her work.

Once this is merged, [P81](https://github.com/zulip/python-zulip-api/pull/81/files) can be closed, since it achieves the same as `embedded bots: Strip @-mention from message.`, but is outdated.

derAnfaenger commented:
Added StateHandler tests. The commit is shown in the wrong position, should go after `Add functional state handler.` Coverage tests will fail before `Add StateHandler tests.`, but I decided against squashing all StateHandler commits, since that felt too cluttered.

timabbott commented:
@derAnfaenger can you submit the StateHandler stuff also in its own PR?  I think it's a relatively small, isolated piece and thus relatively easy to merge.

derAnfaenger commented:
Adding StateHandler in #7043 .

derAnfaenger commented:
Added two commits to test the embedded bot message flow.

This also includes a new helper method `create_test_bot` for `ZulipTestCase`. I included it in the general test class because it could simplify bot creation in other suites like `test_service_bot_system`.

timabbott commented on zerver/lib/events.py:
```diff
@@ -157,6 +158,11 @@ def fetch_initial_state_data(user_profile, event_types, queue_id,
+    # new entries for EMBEDDED_BOTS can only be added directly in the codebase.
+    if want('realm_embedded_bots'):
+        state['realm_embedded_bots'] = list(bot.name for bot in EMBEDDED_BOTS)
+
```
In terms of things that are easy to merge now without adding the feature, this section `realm_embedded_bots` code could easily be submitted as its own PR and we could just merge it.  Feel free to do that if you like.

(Just thinking about what pieces we can merge without turning the feature on)

derAnfaenger replied:
I split out a respective commit and moved it to the beginning of the PR, so you can easily merge it. Out of interest, does it make any difference for you if the commit is in the beginning or in the middle?

timabbott replied:
Great, merging that preliminary as soon as the tests pass.  

It doesn't have to be at the beginning, but it save a bit of time when rebasing, and it's also a way to communicate which things can be merged first :)

timabbott commented:
You should rebase -- it'll auto-drop those already-merged linter commits.

timabbott commented on zerver/views/users.py:
```diff
@@ -261,6 +262,11 @@ def add_bot_backend(request, user_profile, full_name_raw=REQ("full_name"), short
+    if bot_type == UserProfile.EMBEDDED_BOT:
+        is_service_valid = any(service_name == embedded_bot_service.name for embedded_bot_service in EMBEDDED_BOTS)
+        if not is_service_valid:
+            return json_error(_("Invalid service name."))
```
I'm not a fan of this `any` constructon.  I would spell it as:

```
if service_name not in [embedded_bot_service.name for embedded_bot_service in EMBEDDED_BOTS]:
    return json_error(_("Invalid embedded bot name."))
```

(I don't think "service" should appear in user-facing strings at all)

Otherwise, this commit looks good.

derAnfaenger replied:
Updated the `any` construction in the commit `users.py: Clean up and shorten bot handling code.` . Is it possible that you were referring to that commit, but made the comment on Abhijeet's original commit?

timabbott replied:
I think probably you should create a squashed/clean up version of this backend commit closer to the bottom of the branch that we can try to merge next.

(Just so you understand strategy, I'm trying to avoid adding the UI until this works how we want it to from a security perspective).

derAnfaenger replied:
> Just so you understand strategy, I'm trying to avoid adding the UI until this works how we want it to from a security perspective.

Yeah, that makes sense, now shortly before the release. Do you already see any specific sec issues or is it more that you haven't got to carefully review the frontend part yet?

timabbott replied:
It's more that backend bugs are where security issues come from.

derAnfaenger replied:
Will refactor then. It would come really convenient if you could merge the commit `embedded bots: Alter tests to accept any class name. ` from #7043, since that's what prevents the tests from passing. It'd be easier for me to judge any mistakes in this PR afterwards.

timabbott replied:
Sounds good.  Just merged that :)

derAnfaenger commented:
I squashed the code style enhacement commits with Abhijeet's commits and reordered everyting (here is the correct order):

```.py
$ git log -n 8 --pretty=oneline
<<<<<< Frontend stuff
7915fc8647413654c866d4ae27b12475fa352562 bot settings: Fetch embedded bots list from page_params.
1ca1893b94106c42fea25463fb0676ebcdaf1f12 Embedded bots: Add UI for adding embedded bots.
<<<<<< Backend stuff, can be merged independent of frontend stuff
05d73462752900934a2bbec5b12d4c7748e439c9 embedded bots: Add message flow tests.
f507c6c961960e1da9400062ca0a8377099ecdd5 Embedded bots: Add tests for backend coverage of embedded bots.
34c0bea76d46ea18f6a1bfd3af5b73e89dfd7aa1 Embedded bots: Add support to add embedded bots from UI.
<<<<< This can be merged easily and independently of the rest
36f9fc062a34016bdf07e917cc2ab1d87b80161c embedded bots: Properly reply to PMs and streams
1271dbe0419f350e9bb9c973d1fa9c4922ba89d0 embedded bots: Strip @-mention from message.
be026c77a1eac62868c298699585e0ad586e83cd bot services: Use call_consume_in_tests.
```

timabbott commented:
I merged a few commits.

timabbott commented:
I merged most of the backend stuff, but added a `settings.EMBEDDED_BOTS_ENABLED` setting (on by default in dev/test; off by default in production) so that we don't have to worry about doing a full security audit before people can use this in production.

timabbott commented:
See https://github.com/zulip/python-zulip-api/issues/138 for my thought on what we should do with the mention extraction stuff.

roberthoenig commented:
Resolved merge conflicts, fixed some linter quirks, added another node test, and re-tested manually. @showell can you review/merge once this passes Travis?

Note that #P138 still remains an issue for embedded bots. I plan on fixing it as soon as this is merged, for both external and embedded bots. For now, this PR provides embedded bots with the same @-mention-strip functionality as external bots.

showell commented:
merged thanks @roberthoenig

roberthoenig commented:
Awesome! Can you do a server deploy so we can see this in effect? (This requires settings.EMBEDDED_BOTS_ENABLED to be set to True, is that the case?)

showell commented:
You can ask Rishi on chat to do a server deploy.  We're probably overdue for one.

roberthoenig commented:
He did one yesterday, but "Embedded bot" is not an option yet in the bot menu, so I assume `settings.EMBEDDED_BOTS_ENABLED` is set to false.
